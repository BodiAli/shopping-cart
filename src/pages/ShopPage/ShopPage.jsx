import { useEffect, useState } from "react";
import { NavLink, Outlet, useOutletContext } from "react-router-dom";
import CartComponent from "../../components/CartComponent/CartComponent";
import Cart from "../../utils/cart";
import styles from "./ShopPage.module.css";

// Icons
import { SiWindows11, SiApplearcade, SiDwavesystems } from "react-icons/si";
import { BiSolidCategory } from "react-icons/bi";
import { CgGames } from "react-icons/cg";
import { MdLaptopMac } from "react-icons/md";
import { IoCartOutline } from "react-icons/io5";
import { VscTerminalLinux } from "react-icons/vsc";
import { FaPlaystation, FaAppStoreIos } from "react-icons/fa";
import { TiVendorAndroid } from "react-icons/ti";

import {
  GiPistolGun,
  GiMountainRoad,
  GiOpenBook,
  GiCarWheel,
  Gi3dGlasses,
  GiChessKnight,
} from "react-icons/gi";
import { FaPuzzlePiece } from "react-icons/fa6";

function ShopPage() {
  const [isAsideOpen, setIsAsideOpen] = useState(false);
  const [isActive, setIsActive] = useState(false);
  const { arr, handleButtonClick } = useOutletContext();
  const [cart, setCart] = arr;

  useEffect(() => {
    document.title = "Shop | GameVault";

    if (isAsideOpen) {
      document.body.style.overflow = "hidden";
    } else {
      document.body.style.overflow = "initial";
    }

    return () => {
      document.title = "GameVault";
    };
  }, [isAsideOpen]);

  function handleViewAside() {
    setIsAsideOpen(!isAsideOpen);
  }

  function handleViewCart() {
    setIsActive(!isActive);
  }

  function handleProductRemove(game) {
    const newCart = new Cart(cart.products);
    newCart.removeProduct(game);
    setCart(newCart);
  }

  function handleInputChange(e, game) {
    const number = parseFloat(e.target.value);
    if (!isNaN(number) && number > 0) {
      const quantity = number;

      setCart(
        new Cart(
          cart.products.map((product) => {
            if (product === game) {
              return {
                ...product,
                quantity: quantity,
                price: product.originalPrice * quantity,
              };
            } else {
              return product;
            }
          })
        )
      );
    }
  }

  function removeProducts() {
    const newCart = new Cart([]);
    setCart(newCart);
    setIsActive(false);
  }

  return (
    <main className={styles.main}>
      <button
        onClick={handleViewAside}
        className={`${styles.viewAsideButton} ${isAsideOpen ? styles.isActive : ""}`}
      >
        <div className={styles.bar}></div>
      </button>
      <aside className={`${styles.aside} ${isAsideOpen ? styles.isOpen : ""}`}>
        <div onClick={handleViewCart} className={styles.viewCart}>
          <h2>
            <IoCartOutline size={"1.5em"} />
            View Cart
          </h2>
          <div data-testid="quantity" className={styles.count}>
            {cart.count}
          </div>
        </div>
        <h2>
          <CgGames size={"1.3em"} />
          Games
        </h2>
        <ul className={styles.navigation}>
          <NavLink to="games">
            <CgGames size={"1.2em"} />
            Games
          </NavLink>
        </ul>

        <h2>
          <SiDwavesystems size={"1.3em"} />
          Platforms
        </h2>
        <ul className={styles.navigation}>
          <NavLink to="windows">
            <SiWindows11 />
            Windows
          </NavLink>

          <NavLink to="mac">
            <MdLaptopMac size={"1.2em"} />
            Mac
          </NavLink>

          <NavLink to="linux">
            <VscTerminalLinux size={"1.2em"} />
            Linux
          </NavLink>

          <NavLink to="playstation">
            <FaPlaystation size={"1.2em"} />
            Playstation
          </NavLink>

          <NavLink to="android">
            <TiVendorAndroid size={"1.2em"} />
            Android
          </NavLink>

          <NavLink to="iphone">
            <FaAppStoreIos size={"1.2em"} />
            iPhone
          </NavLink>
        </ul>
        <h2>
          <BiSolidCategory size={"1.3em"} />
          Genres
        </h2>
        <ul className={styles.navigation}>
          <NavLink to="action">
            <GiPistolGun size={"1.2em"} />
            Action
          </NavLink>

          <NavLink to="adventure">
            <GiMountainRoad size={"1.2em"} />
            Adventure
          </NavLink>

          <NavLink to="education">
            <GiOpenBook size={"1.2em"} />
            Educational
          </NavLink>

          <NavLink to="puzzle">
            <FaPuzzlePiece size={"1.2em"} />
            Puzzle
          </NavLink>

          <NavLink to="arcade">
            <SiApplearcade size={"1.2em"} />
            Arcade
          </NavLink>

          <NavLink to="racing">
            <GiCarWheel size={"1.2em"} />
            Racing
          </NavLink>

          <NavLink to="simulation">
            <Gi3dGlasses size={"1.2em"} />
            Simulation
          </NavLink>

          <NavLink to="strategy">
            <GiChessKnight size={"1.2em"} />
            Strategy
          </NavLink>
        </ul>
      </aside>
      <CartComponent
        handleCartVisibility={handleViewCart}
        cart={cart}
        handleInputChange={handleInputChange}
        isActive={isActive}
        handleProductRemove={handleProductRemove}
        removeProducts={removeProducts}
      />
      <Outlet context={{ arr, handleButtonClick }} />
    </main>
  );
}

export default ShopPage;
