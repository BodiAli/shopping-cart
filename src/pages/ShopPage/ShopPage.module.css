.main {
  background-color: var(--green);
  display: grid;
  grid-template-columns: 1fr 6fr;
  position: relative;
  overflow: hidden;
}

.aside {
  background-color: var(--gray);
  border-radius: 5px;
  max-height: calc(100vh - 88px);
  overflow-y: auto;
  padding: 1.5rem 0;
  scrollbar-width: thin;
  scrollbar-color: var(--primary) var(--gray);
}

.aside > h2 {
  font-size: 1.2rem;
  font-weight: 600;
  margin-bottom: 10px;
  padding: 0 0.5rem;
  text-shadow: 0 0 1px rgba(0, 0, 0, 0.775);
}

.navigation {
  display: flex;
  flex-direction: column;
  gap: 30px;
}

.navigation:not(:last-child) {
  margin-bottom: 50px;
}

.navigation a {
  display: flex;
  align-items: center;
  gap: 5px;
  color: var(--navText);
  text-shadow: 0 0 10px rgba(0, 0, 0, 0.775);
  text-decoration: none;
  font-weight: 500;
  padding: 1rem 0.5rem;
  background-color: var(--darkerGray);
}

.navigation :global(a.active) {
  background-color: var(--darkerGreen);
}

.navigation a:hover {
  filter: brightness(1.1);
}

.viewCart {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.8rem 0.5rem;
  margin-bottom: 50px;
  background-color: var(--darkerGray);
  cursor: pointer;
  border: 5px outset white;
}

.viewCart:hover {
  filter: brightness(1.1);
}

.viewCart:active {
  filter: brightness(0.9);
  border: 3px solid white;
}

.viewCart > h2 {
  font-size: 1.1rem;
  font-weight: 600;
  display: flex;
  align-items: center;
  gap: 5px;
}

.count {
  background-color: white;
  border-radius: 50%;
  padding: 2px 5px;
  width: 2ch;
  display: flex;
  justify-content: center;
  align-items: center;
}
