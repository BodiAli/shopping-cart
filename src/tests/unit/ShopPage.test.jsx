import { render, screen } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import { vi } from "vitest";
import ShopPage from "../../pages/ShopPage/ShopPage";
import Cart from "../../utils/cart";

const cart = new Cart([{ title: "Call of Duty" }]);

// Mocking useOutletContext to return an array because it returns null and causes an error
vi.mock("react-router-dom", async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    useOutletContext: vi.fn(() => {
      return { arr: [cart] };
    }),
  };
});

// Test if it renders the CartComponent with cart object
vi.mock("../../components/CartComponent/CartComponent", () => {
  return {
    default: vi.fn(({ cart }) => <p>{cart.products[0].title}</p>),
  };
});

const router = createMemoryRouter([
  {
    path: "/",
    element: <ShopPage />,
  },
]);

const wrapper = () => <RouterProvider router={router} />;

describe("Shop page", () => {
  test("should render correct elements", () => {
    const { container } = render(<ShopPage />, { wrapper });
    expect(container).toMatchSnapshot();
  });

  test("should render games nav", () => {
    render(<ShopPage />, { wrapper });
    expect(screen.getByRole("heading", { name: /games/i })).toBeInTheDocument();

    const gamesNav = screen.getByRole("link", { name: /games/i });

    expect(gamesNav).toBeInTheDocument();
    expect(gamesNav).toHaveAttribute("href", "/games");
  });

  test("should render platforms navigation", () => {
    render(<ShopPage />, { wrapper });
    expect(screen.getByRole("heading", { name: /platforms/i })).toBeInTheDocument();

    const windowsNav = screen.getByRole("link", { name: /windows/i });
    const macNav = screen.getByRole("link", { name: /mac/i });
    const linuxNav = screen.getByRole("link", { name: /linux/i });
    const androidNav = screen.getByRole("link", { name: /android/i });
    const iPhoneNav = screen.getByRole("link", { name: /iphone/i });

    expect(windowsNav).toBeInTheDocument();
    expect(macNav).toBeInTheDocument();
    expect(linuxNav).toBeInTheDocument();
    expect(androidNav).toBeInTheDocument();
    expect(iPhoneNav).toBeInTheDocument();

    expect(windowsNav).toHaveAttribute("href", "/windows");
    expect(macNav).toHaveAttribute("href", "/mac");
    expect(linuxNav).toHaveAttribute("href", "/linux");
    expect(androidNav).toHaveAttribute("href", "/android");
    expect(iPhoneNav).toHaveAttribute("href", "/iphone");
  });

  test("should render genres navigation", () => {
    render(<ShopPage />, { wrapper });
    expect(screen.getByRole("heading", { name: /genres/i })).toBeInTheDocument();

    const actionNav = screen.getByRole("link", { name: /action/i });
    const adventureNav = screen.getByRole("link", { name: /adventure/i });
    const educationNav = screen.getByRole("link", { name: /education/i });
    const arcadeNav = screen.getByRole("link", { name: /arcade/i });
    const racingNav = screen.getByRole("link", { name: /racing/i });
    const simulationNav = screen.getByRole("link", { name: /simulation/i });
    const puzzleNav = screen.getByRole("link", { name: /puzzle/i });
    const strategyNav = screen.getByRole("link", { name: /strategy/i });

    expect(actionNav).toBeInTheDocument();
    expect(adventureNav).toBeInTheDocument();
    expect(educationNav).toBeInTheDocument();
    expect(arcadeNav).toBeInTheDocument();
    expect(racingNav).toBeInTheDocument();
    expect(simulationNav).toBeInTheDocument();
    expect(puzzleNav).toBeInTheDocument();
    expect(strategyNav).toBeInTheDocument();

    expect(actionNav).toHaveAttribute("href", "/action");
    expect(adventureNav).toHaveAttribute("href", "/adventure");
    expect(educationNav).toHaveAttribute("href", "/education");
    expect(arcadeNav).toHaveAttribute("href", "/arcade");
    expect(racingNav).toHaveAttribute("href", "/racing");
    expect(simulationNav).toHaveAttribute("href", "/simulation");
    expect(puzzleNav).toHaveAttribute("href", "/puzzle");
    expect(strategyNav).toHaveAttribute("href", "/strategy");
  });

  test("should render an element to view cart", () => {
    render(<ShopPage />, { wrapper });
    expect(screen.getByRole("heading", { name: /view cart/i })).toBeInTheDocument();
  });

  test("should render number of cart items in the cart", async () => {
    const { unmount } = render(<ShopPage />, { wrapper });
    expect(screen.getByTestId("quantity")).toHaveTextContent(1);

    unmount();
    cart.addProduct({ title: "Fifa" });
    cart.addProduct({ title: "Minecraft" });

    render(<ShopPage />, { wrapper });
    expect(screen.getByTestId("quantity")).toHaveTextContent(3);
  });
});
