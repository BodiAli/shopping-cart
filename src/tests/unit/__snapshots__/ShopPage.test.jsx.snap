// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop page > should render correct elements 1`] = `
<div>
  <main
    class="_main_c02aaf"
  >
    <aside
      class="_aside_c02aaf"
    >
      <h2>
        Games
      </h2>
      <ul
        class="_navigation_c02aaf"
      >
        <a
          class=""
          href="/games"
        >
          Games
        </a>
      </ul>
      <h2>
        Platforms
      </h2>
      <ul
        class="_navigation_c02aaf"
      >
        <a
          class=""
          href="/windows"
        >
          Windows
        </a>
        <a
          class=""
          href="/mac"
        >
          Mac
        </a>
        <a
          class=""
          href="/linux"
        >
          Linux
        </a>
      </ul>
      <h2>
        Genres
      </h2>
      <ul
        class="_navigation_c02aaf"
      >
        <a
          class=""
          href="/action"
        >
          Action
        </a>
        <a
          class=""
          href="/adventure"
        >
          Adventure
        </a>
        <a
          class=""
          href="/education"
        >
          Education
        </a>
        <a
          class=""
          href="/arcade"
        >
          Arcade
        </a>
        <a
          class=""
          href="/racing"
        >
          Racing
        </a>
        <a
          class=""
          href="/simulation"
        >
          Simulation
        </a>
        <a
          class=""
          href="/sports"
        >
          Sports
        </a>
        <a
          class=""
          href="/strategy"
        >
          Strategy
        </a>
        <a
          class=""
          href="/role-playing"
        >
          Role Playing
        </a>
      </ul>
    </aside>
  </main>
</div>
`;
